hal/watchdog_api.h

```diff
@@ -0,0 +1,149 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_WATCHDOG_API_H
+#define MBED_WATCHDOG_API_H
+
+#if DEVICE_WATCHDOG
+
+#include <stdbool.h>
+#include <stdint.h>
+
+/**
+ * \defgroup hal_watchdog Watchdog HAL API
+ * @{
+ */
+
+/** \file watchdog_api.h
+ *
+ * This module provides platform independent access to the system watchdog timer
+ * which is an embedded peripheral that will reset the system in the case of
+ * system failures or malfunctions.
+ *
+ * The watchdog timer initialises a system timer with a time period specified in
+ * the configuration. This timer counts down and triggers a system reset when it
+ * wraps. To prevent the system reset the timer must be continually
+ * kicked/refreshed by calling hal_watchdog_kick which will reset the countdown
+ * to the user specified reset value.
+ *
+ * The Watchdog timer must continue to operate in low power modes. It
+ * must count down and trigger a reset from within both sleep and deep sleep
+ * modes unless the chip is woken to refresh the timer.
+ */
+
+typedef struct {
+    /**
+    * Refresh value for the watchdog in milliseconds. The maximum value of this
+    * setting is platform dependent, to find the maximum value for the current
+    * platform call hal_watchdog_get_features() and check the timeout value
+    * member. The minimum valid value for this setting is 1, attempting to
+    * initialise the watchdog with a timeout of 0ms will return
+    * WATCHDOG_STATUS_INVALID_ARGUMENT.
+    */
+    uint32_t timeout_ms;
+} watchdog_config_t;
+
+
+typedef struct {
+    /**
+     * Maximum timeout value for the watchdog in milliseconds.
+     */
+    uint32_t max_timeout;
+    /**
+     * Watchdog configuration can be updated after the watchdog has been started
+     */
+    bool update_config;
+    /**
+     * Watchdog can be stopped after it is started without a reset
+     */
+    bool disable_watchdog;
+} watchdog_features_t;
+
+
+typedef enum {
```

Comment:
Maybe we can use Mbed-OS standardized error coding here, which already defines a code for each of these.